hostapd: add support for subject validation

swg0101 opened the pull request:
hostapd: add support for subject validation

The wpa_supplicant supports certificate subject validation via the
subject match(2) and altsubject_match(2) fields. domain_match(2) and 
domain_suffix_match(2) fields are also supported for advanced matches. 
This validation is especially important when connecting to access 
points that use PAP as the Phase 2 authentication type. Without proper 
validation, the user's password can be transmitted to a rogue access 
point in plaintext without the user's knowledge. Most organizations 
already require these attributes to be included to ensure that the 
connection from the STA and the AP is secure. Includes LuCI changes via 
openwrt/luci#3444.

From the documentation:

subject_match - Constraint for server certificate subject. This substring
is matched against the subject of the authentication server certificate.
If this string is set, the server sertificate is only accepted if it
contains this string in the subject. The subject string is in following
format: /C=US/ST=CA/L=San Francisco/CN=Test AS/emailAddress=as
.example.com

subject_match2 - Constraint for server certificate subject. This field is
like subject_match, but used for phase 2 (inside EAP-TTLS/PEAP/FAST
tunnel) authentication.

altsubject_match - Constraint for server certificate alt. subject.
Semicolon separated string of entries to be matched against the
alternative subject name of the authentication server certificate. If
this string is set, the server sertificate is only accepted if it
contains one of the entries in an alternative subject name extension.
altSubjectName string is in following format: TYPE:VALUE Example:
EMAIL:server@example.com Example:
DNS:server.example.com;DNS:server2.example.com Following types are
supported: EMAIL, DNS, URI

altsubject_match2 - Constraint for server certificate alt. subject. This
field is like altsubject_match, but used for phase 2 (inside
EAP-TTLS/PEAP/FAST tunnel) authentication.

domain_match - Constraint for server domain name. If set, this FQDN is 
used as a full match requirement for the
server certificate in SubjectAltName dNSName element(s). If a
matching dNSName is found, this constraint is met. If no dNSName
values are present, this constraint is matched against SubjectName CN
using same full match comparison. This behavior is similar to
domain_suffix_match, but has the requirement of a full match, i.e.,
no subdomains or wildcard matches are allowed. Case-insensitive
comparison is used, so "Example.com" matches "example.com", but would
not match "test.Example.com". More than one match string can be 
provided by using semicolons to
separate the strings (e.g., example.org;example.com). When multiple
strings are specified, a match with any one of the values is considered
a sufficient match for the certificate, i.e., the conditions are ORed
together.

domain_match2 - Constraint for server domain name. This field is like 
domain_match, but used for phase 2 (inside EAP-TTLS/PEAP/FAST tunnel) 
authentication.

domain_suffix_match - Constraint for server domain name. If set, this 
FQDN is used as a suffix match requirement for the AAA server 
certificate in SubjectAltName dNSName element(s). If a matching dNSName 
is found, this constraint is met. If no dNSName values are present, 
this constraint is matched against SubjectName CN using same suffix 
match comparison. Suffix match here means that the host/domain name is 
compared one label at a time starting from the top-level domain and all 
the labels in domain_suffix_match shall be included in the certificate. 
The certificate may include additional sub-level labels in addition to 
the required labels. More than one match string can be provided by using 
semicolons to separate the strings (e.g., example.org;example.com). 
When multiple strings are specified, a match with any one of the values 
is considered a sufficient match for the certificate, i.e., the 
conditions are ORed together. For example, 
domain_suffix_match=example.com would match test.example.com but would 
not match test-example.com. This field is like domain_match, but used 
for phase 2 (inside EAP-TTLS/PEAP/FAST tunnel) authentication.

domain_suffix_match2 - Constraint for server domain name. This field is 
like domain_suffix_match, but used for phase 2 (inside 
EAP-TTLS/PEAP/FAST tunnel) authentication.

Signed-off-by: David Lam <david@thedavid.net>

charlessuh commented:
Hi, there.

I've also been looking into this and it looks like you beat me to the punch.

Would you be interested in expanding the scope of your PR to make this more thorough by supporting wildcards at the beginning of the subject name and also supporting multiple values? For example, I was working along the lines of the following:

```
hostapd_append_cert_match() {
    local network_data "$1"

    json_get_vars subject_match
    [ -n "$subject_match" ] && append network_data "subject_match=\"$subject_match\"" "$N$T"

    local alist= a
    json_get_values altsubject_list altsubject_match
    for a in $altsubject_list; do
        append alist "$a" ";"
    done
                
    [ -n "$alist" ] && append network_data "altsubject_match=\"$alist\"" "$N$T"

    local dslist= ds
    json_get_values domain_suffix_list domain_suffix_match
    for ds in $domain_suffix_list; do
        append dslist "$ds" ";"
    done

    [ -n "$dslist" ] && append network_data "domain_suffix_match=\"$dslist\"" "$N$T"

    local dlist= d
    json_get_values domain_list domain_match
    for d in $domain_list; do
        append dlist "$d" ";"
    done

    [ -n "$dlist" ] && append network_data "domain_match=\"$dlist\"" "$N$T"
}
```

If not, that's totally fine as well -- I'll just wait for your PR to get potentially merged to avoid any stomping of your changes. Thanks!

swg0101 commented:
I can revise my change to accept lists instead of static values. That would allow LuCI to accept multiple values more easily. As far as domain_match is concerned, this is mainly redundant since altsubject_match does the exact same thing. (e.g. DNS:radius.example.com in altsubject_match == radius.example.com in a domain_match).

What do you think?

charlessuh commented:
> I can revise my change to accept lists instead of static values. That would allow LuCI to accept multiple values more easily. As far as domain_match is concerned, this is mainly redundant since altsubject_match does the exact same thing. (e.g. DNS:radius.example.com in altsubject_match == radius.example.com in a domain_match). domain_suffix_match could be a bit more useful since there is no other way I know of to match suffixes on a SAN (subject_match only uses a substring match against the Subject field).
> 
> What do you think?

Cool, thanks!

Yeah, I really only need `domain_suffix_match` for my use case, and just added `domain_match` in that example to be thorough.

(Note it looks like `domain_match` also looks at the common name. It's possible there are people with self-signed certs that are only using the common name, as I recall that the `openssl` command line tool might not prompt by default for extensions like subject alt name. But feel free to drop it as this is something that I don't need.)

swg0101 commented:
Yeah, I saw that as well. But then since that behavior is somewhat ambiguous, I think it's probably better to do it via subject_match because the values are printed out directly on the log if you read it via the console using logread -f.

That said, give me a day or two and I will get this revised. Thanks!

swg0101 commented:
@charlessuh I actually was looking at domain_match again and it looks like there may be some edge cases when it may be useful since it allows for multiple exact matches whereas subject_match only allows for one substring match. I have updated the commits to incorporate these changes. Can you take a look and see what you think?

charlessuh commented:
Awesome!

@nbd168 Are you the maintainer of openwrt's hostapd package? If so, would you be able to review for inclusion into `master`? :)

jow- commented:
@swg0101 - that looks good to go to me. Can you please increase the PKG_REVISION in the Makefile as well?

swg0101 committed:
hostapd: add support for subject validation

The wpa_supplicant supports certificate subject validation via the
subject match(2) and altsubject_match(2) fields. domain_match(2) and
domain_suffix_match(2) fields are also supported for advanced matches.
This validation is especially important when connecting to access
points that use PAP as the Phase 2 authentication type. Without proper
validation, the user's password can be transmitted to a rogue access
point in plaintext without the user's knowledge. Most organizations
already require these attributes to be included to ensure that the
connection from the STA and the AP is secure. Includes LuCI changes via
openwrt/luci#3444.

From the documentation:

subject_match - Constraint for server certificate subject. This substring
is matched against the subject of the authentication server certificate.
If this string is set, the server sertificate is only accepted if it
contains this string in the subject. The subject string is in following
format: /C=US/ST=CA/L=San Francisco/CN=Test AS/emailAddress=as
.example.com

subject_match2 - Constraint for server certificate subject. This field is
like subject_match, but used for phase 2 (inside EAP-TTLS/PEAP/FAST
tunnel) authentication.

altsubject_match - Constraint for server certificate alt. subject.
Semicolon separated string of entries to be matched against the
alternative subject name of the authentication server certificate. If
this string is set, the server sertificate is only accepted if it
contains one of the entries in an alternative subject name extension.
altSubjectName string is in following format: TYPE:VALUE Example:
EMAIL:server@example.com Example:
DNS:server.example.com;DNS:server2.example.com Following types are
supported: EMAIL, DNS, URI

altsubject_match2 - Constraint for server certificate alt. subject. This
field is like altsubject_match, but used for phase 2 (inside
EAP-TTLS/PEAP/FAST tunnel) authentication.

domain_match - Constraint for server domain name. If set, this FQDN is
used as a full match requirement for the
server certificate in SubjectAltName dNSName element(s). If a
matching dNSName is found, this constraint is met. If no dNSName
values are present, this constraint is matched against SubjectName CN
using same full match comparison. This behavior is similar to
domain_suffix_match, but has the requirement of a full match, i.e.,
no subdomains or wildcard matches are allowed. Case-insensitive
comparison is used, so "Example.com" matches "example.com", but would
not match "test.Example.com". More than one match string can be
provided by using semicolons to
separate the strings (e.g., example.org;example.com). When multiple
strings are specified, a match with any one of the values is considered
a sufficient match for the certificate, i.e., the conditions are ORed
together.

domain_match2 - Constraint for server domain name. This field is like
domain_match, but used for phase 2 (inside EAP-TTLS/PEAP/FAST tunnel)
authentication.

domain_suffix_match - Constraint for server domain name. If set, this
FQDN is used as a suffix match requirement for the AAA server
certificate in SubjectAltName dNSName element(s). If a matching dNSName
is found, this constraint is met. If no dNSName values are present,
this constraint is matched against SubjectName CN using same suffix
match comparison. Suffix match here means that the host/domain name is
compared one label at a time starting from the top-level domain and all
the labels in domain_suffix_match shall be included in the certificate.
The certificate may include additional sub-level labels in addition to
the required labels. More than one match string can be provided by using
semicolons to separate the strings (e.g., example.org;example.com).
When multiple strings are specified, a match with any one of the values
is considered a sufficient match for the certificate, i.e., the
conditions are ORed together. For example,
domain_suffix_match=example.com would match test.example.com but would
not match test-example.com. This field is like domain_match, but used
for phase 2 (inside EAP-TTLS/PEAP/FAST tunnel) authentication.

domain_suffix_match2 - Constraint for server domain name. This field is
like domain_suffix_match, but used for phase 2 (inside
EAP-TTLS/PEAP/FAST tunnel) authentication.

Signed-off-by: David Lam <david@thedavid.net>

swg0101 commented:
Done!

jow- commented:
Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/jow.git
